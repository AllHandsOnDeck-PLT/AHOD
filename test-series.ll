; ModuleID = 'AHOD'
source_filename = "AHOD"

@str = global [4 x i8] c"%s\0A\00"
@str.1 = global [4 x i8] c"%d\0A\00"
@str.2 = global [4 x i8] c"%g\0A\00"
@str.3 = global [4 x i8] c"%d\0A\00"
@a = global { i32*, i32* } zeroinitializer

define i32 @main() {
entry:
  %new_series_ptr = alloca { i32*, i32* }
  %series_size_ptr = getelementptr inbounds { i32*, i32* }, { i32*, i32* }* %new_series_ptr, i32 0, i32 0
  %series_size = alloca i32
  store i32 0, i32* %series_size
  store i32* %series_size, i32** %series_size_ptr
  %series.arry = getelementptr inbounds { i32*, i32* }, { i32*, i32* }* %new_series_ptr, i32 0, i32 1
  %p = alloca i32, i32 1028
  store i32* %p, i32** %series.arry
  call void @series_addint({ i32*, i32* }* %new_series_ptr, i32 1)
  call void @series_addint({ i32*, i32* }* %new_series_ptr, i32 3)
  %new_series = load { i32*, i32* }, { i32*, i32* }* %new_series_ptr
  store { i32*, i32* } %new_series, { i32*, i32* }* @a
  %series_get = call i32 @series_getint({ i32*, i32* }* @a, i32 1)
  %printf = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @str.1, i32 0, i32 0), i32 %series_get)
  ret i32 0
}

declare i32 @printf(i8*, ...)

define void @series_addbool({ i32*, i1* }*, i1) {
entry:
  %series_ptr_alloc = alloca { i32*, i1* }*
  store { i32*, i1* }* %0, { i32*, i1* }** %series_ptr_alloc
  %val_alloc = alloca i1
  store i1 %1, i1* %val_alloc
  %series_load = load { i32*, i1* }*, { i32*, i1* }** %series_ptr_alloc
  %series_array_ptr = getelementptr inbounds { i32*, i1* }, { i32*, i1* }* %series_load, i32 0, i32 1
  %series_array_load = load i1*, i1** %series_array_ptr
  %series_size_ptr_ptr = getelementptr inbounds { i32*, i1* }, { i32*, i1* }* %series_load, i32 0, i32 0
  %series_size_ptr = load i32*, i32** %series_size_ptr_ptr
  %series_size = load i32, i32* %series_size_ptr
  %series_arry_next_element_ptr = getelementptr i1, i1* %series_array_load, i32 %series_size
  %inc_size = add i32 %series_size, 1
  store i32 %inc_size, i32* %series_size_ptr
  %val = load i1, i1* %val_alloc
  store i1 %val, i1* %series_arry_next_element_ptr
  ret void
}

define void @series_addint({ i32*, i32* }*, i32) {
entry:
  %series_ptr_alloc = alloca { i32*, i32* }*
  store { i32*, i32* }* %0, { i32*, i32* }** %series_ptr_alloc
  %val_alloc = alloca i32
  store i32 %1, i32* %val_alloc
  %series_load = load { i32*, i32* }*, { i32*, i32* }** %series_ptr_alloc
  %series_array_ptr = getelementptr inbounds { i32*, i32* }, { i32*, i32* }* %series_load, i32 0, i32 1
  %series_array_load = load i32*, i32** %series_array_ptr
  %series_size_ptr_ptr = getelementptr inbounds { i32*, i32* }, { i32*, i32* }* %series_load, i32 0, i32 0
  %series_size_ptr = load i32*, i32** %series_size_ptr_ptr
  %series_size = load i32, i32* %series_size_ptr
  %series_arry_next_element_ptr = getelementptr i32, i32* %series_array_load, i32 %series_size
  %inc_size = add i32 %series_size, 1
  store i32 %inc_size, i32* %series_size_ptr
  %val = load i32, i32* %val_alloc
  store i32 %val, i32* %series_arry_next_element_ptr
  ret void
}

define void @series_addfloat({ i32*, double* }*, double) {
entry:
  %series_ptr_alloc = alloca { i32*, double* }*
  store { i32*, double* }* %0, { i32*, double* }** %series_ptr_alloc
  %val_alloc = alloca double
  store double %1, double* %val_alloc
  %series_load = load { i32*, double* }*, { i32*, double* }** %series_ptr_alloc
  %series_array_ptr = getelementptr inbounds { i32*, double* }, { i32*, double* }* %series_load, i32 0, i32 1
  %series_array_load = load double*, double** %series_array_ptr
  %series_size_ptr_ptr = getelementptr inbounds { i32*, double* }, { i32*, double* }* %series_load, i32 0, i32 0
  %series_size_ptr = load i32*, i32** %series_size_ptr_ptr
  %series_size = load i32, i32* %series_size_ptr
  %series_arry_next_element_ptr = getelementptr double, double* %series_array_load, i32 %series_size
  %inc_size = add i32 %series_size, 1
  store i32 %inc_size, i32* %series_size_ptr
  %val = load double, double* %val_alloc
  store double %val, double* %series_arry_next_element_ptr
  ret void
}

define void @series_addstr({ i32*, i8** }*, i8*) {
entry:
  %series_ptr_alloc = alloca { i32*, i8** }*
  store { i32*, i8** }* %0, { i32*, i8** }** %series_ptr_alloc
  %val_alloc = alloca i8*
  store i8* %1, i8** %val_alloc
  %series_load = load { i32*, i8** }*, { i32*, i8** }** %series_ptr_alloc
  %series_array_ptr = getelementptr inbounds { i32*, i8** }, { i32*, i8** }* %series_load, i32 0, i32 1
  %series_array_load = load i8**, i8*** %series_array_ptr
  %series_size_ptr_ptr = getelementptr inbounds { i32*, i8** }, { i32*, i8** }* %series_load, i32 0, i32 0
  %series_size_ptr = load i32*, i32** %series_size_ptr_ptr
  %series_size = load i32, i32* %series_size_ptr
  %series_arry_next_element_ptr = getelementptr i8*, i8** %series_array_load, i32 %series_size
  %inc_size = add i32 %series_size, 1
  store i32 %inc_size, i32* %series_size_ptr
  %val = load i8*, i8** %val_alloc
  store i8* %val, i8** %series_arry_next_element_ptr
  ret void
}

define i1 @series_getbool({ i32*, i1* }*, i32) {
entry:
  %series_ptr_alloc = alloca { i32*, i1* }*
  store { i32*, i1* }* %0, { i32*, i1* }** %series_ptr_alloc
  %idx_alloc = alloca i32
  store i32 %1, i32* %idx_alloc
  %series_load = load { i32*, i1* }*, { i32*, i1* }** %series_ptr_alloc
  %series_array_ptr = getelementptr inbounds { i32*, i1* }, { i32*, i1* }* %series_load, i32 0, i32 1
  %array_load = load i1*, i1** %series_array_ptr
  %idx_load = load i32, i32* %idx_alloc
  %series_arry_element_ptr = getelementptr i1, i1* %array_load, i32 %idx_load
  %series_array_element_ptr = load i1, i1* %series_arry_element_ptr
  ret i1 %series_array_element_ptr
}

define i32 @series_getint({ i32*, i32* }*, i32) {
entry:
  %series_ptr_alloc = alloca { i32*, i32* }*
  store { i32*, i32* }* %0, { i32*, i32* }** %series_ptr_alloc
  %idx_alloc = alloca i32
  store i32 %1, i32* %idx_alloc
  %series_load = load { i32*, i32* }*, { i32*, i32* }** %series_ptr_alloc
  %series_array_ptr = getelementptr inbounds { i32*, i32* }, { i32*, i32* }* %series_load, i32 0, i32 1
  %array_load = load i32*, i32** %series_array_ptr
  %idx_load = load i32, i32* %idx_alloc
  %series_arry_element_ptr = getelementptr i32, i32* %array_load, i32 %idx_load
  %series_array_element_ptr = load i32, i32* %series_arry_element_ptr
  ret i32 %series_array_element_ptr
}

define double @series_getfloat({ i32*, double* }*, i32) {
entry:
  %series_ptr_alloc = alloca { i32*, double* }*
  store { i32*, double* }* %0, { i32*, double* }** %series_ptr_alloc
  %idx_alloc = alloca i32
  store i32 %1, i32* %idx_alloc
  %series_load = load { i32*, double* }*, { i32*, double* }** %series_ptr_alloc
  %series_array_ptr = getelementptr inbounds { i32*, double* }, { i32*, double* }* %series_load, i32 0, i32 1
  %array_load = load double*, double** %series_array_ptr
  %idx_load = load i32, i32* %idx_alloc
  %series_arry_element_ptr = getelementptr double, double* %array_load, i32 %idx_load
  %series_array_element_ptr = load double, double* %series_arry_element_ptr
  ret double %series_array_element_ptr
}

define i8* @series_getstr({ i32*, i8** }*, i32) {
entry:
  %series_ptr_alloc = alloca { i32*, i8** }*
  store { i32*, i8** }* %0, { i32*, i8** }** %series_ptr_alloc
  %idx_alloc = alloca i32
  store i32 %1, i32* %idx_alloc
  %series_load = load { i32*, i8** }*, { i32*, i8** }** %series_ptr_alloc
  %series_array_ptr = getelementptr inbounds { i32*, i8** }, { i32*, i8** }* %series_load, i32 0, i32 1
  %array_load = load i8**, i8*** %series_array_ptr
  %idx_load = load i32, i32* %idx_alloc
  %series_arry_element_ptr = getelementptr i8*, i8** %array_load, i32 %idx_load
  %series_array_element_ptr = load i8*, i8** %series_arry_element_ptr
  ret i8* %series_array_element_ptr
}
