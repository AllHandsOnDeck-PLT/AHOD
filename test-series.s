	.text
	.file	"AHOD"
	.globl	list_pushbool           # -- Begin function list_pushbool
	.p2align	4, 0x90
	.type	list_pushbool,@function
list_pushbool:                          # @list_pushbool
	.cfi_startproc
# %bb.0:                                # %entry
	movq	%rdi, -8(%rsp)
	andl	$1, %esi
	movb	%sil, -9(%rsp)
	movq	(%rdi), %rax
	movq	8(%rdi), %rcx
	movslq	(%rax), %rdx
	leal	1(%rdx), %esi
	movl	%esi, (%rax)
	movb	-9(%rsp), %al
	movb	%al, (%rcx,%rdx)
	retq
.Lfunc_end0:
	.size	list_pushbool, .Lfunc_end0-list_pushbool
	.cfi_endproc
                                        # -- End function
	.globl	list_pushint            # -- Begin function list_pushint
	.p2align	4, 0x90
	.type	list_pushint,@function
list_pushint:                           # @list_pushint
	.cfi_startproc
# %bb.0:                                # %entry
	movq	%rdi, -8(%rsp)
	movl	%esi, -12(%rsp)
	movq	(%rdi), %rax
	movq	8(%rdi), %rcx
	movslq	(%rax), %rdx
	leal	1(%rdx), %esi
	movl	%esi, (%rax)
	movl	-12(%rsp), %eax
	movl	%eax, (%rcx,%rdx,4)
	retq
.Lfunc_end1:
	.size	list_pushint, .Lfunc_end1-list_pushint
	.cfi_endproc
                                        # -- End function
	.globl	list_pushfloat          # -- Begin function list_pushfloat
	.p2align	4, 0x90
	.type	list_pushfloat,@function
list_pushfloat:                         # @list_pushfloat
	.cfi_startproc
# %bb.0:                                # %entry
	movq	%rdi, -8(%rsp)
	movsd	%xmm0, -16(%rsp)
	movq	(%rdi), %rax
	movq	8(%rdi), %rcx
	movslq	(%rax), %rdx
	leal	1(%rdx), %esi
	movl	%esi, (%rax)
	movsd	-16(%rsp), %xmm0        # xmm0 = mem[0],zero
	movsd	%xmm0, (%rcx,%rdx,8)
	retq
.Lfunc_end2:
	.size	list_pushfloat, .Lfunc_end2-list_pushfloat
	.cfi_endproc
                                        # -- End function
	.globl	list_pushstr            # -- Begin function list_pushstr
	.p2align	4, 0x90
	.type	list_pushstr,@function
list_pushstr:                           # @list_pushstr
	.cfi_startproc
# %bb.0:                                # %entry
	movq	%rdi, -8(%rsp)
	movq	%rsi, -16(%rsp)
	movq	(%rdi), %rax
	movq	8(%rdi), %rcx
	movslq	(%rax), %rdx
	leal	1(%rdx), %esi
	movl	%esi, (%rax)
	movq	-16(%rsp), %rax
	movq	%rax, (%rcx,%rdx,8)
	retq
.Lfunc_end3:
	.size	list_pushstr, .Lfunc_end3-list_pushstr
	.cfi_endproc
                                        # -- End function
	.globl	main                    # -- Begin function main
	.p2align	4, 0x90
	.type	main,@function
main:                                   # @main
	.cfi_startproc
# %bb.0:                                # %entry
	subq	$4136, %rsp             # imm = 0x1028
	.cfi_def_cfa_offset 4144
	movq	str.1@GOTPCREL(%rip), %rdi
	movl	$1, %esi
	xorl	%eax, %eax
	callq	printf@PLT
	movl	$0, 4(%rsp)
	leaq	4(%rsp), %rax
	movq	%rax, 8(%rsp)
	leaq	24(%rsp), %rax
	movq	%rax, 16(%rsp)
	leaq	8(%rsp), %rdi
	movl	$1, %esi
	callq	list_pushint@PLT
	movq	8(%rsp), %rsi
	movq	16(%rsp), %rdx
	movq	str@GOTPCREL(%rip), %rdi
	xorl	%eax, %eax
	callq	printf@PLT
	xorl	%eax, %eax
	addq	$4136, %rsp             # imm = 0x1028
	retq
.Lfunc_end4:
	.size	main, .Lfunc_end4-main
	.cfi_endproc
                                        # -- End function
	.type	str,@object             # @str
	.data
	.globl	str
str:
	.asciz	"%s\n"
	.size	str, 4

	.type	str.1,@object           # @str.1
	.globl	str.1
str.1:
	.asciz	"%d\n"
	.size	str.1, 4

	.type	str.2,@object           # @str.2
	.globl	str.2
str.2:
	.asciz	"%g\n"
	.size	str.2, 4

	.type	str.3,@object           # @str.3
	.globl	str.3
str.3:
	.asciz	"%d\n"
	.size	str.3, 4


	.section	".note.GNU-stack","",@progbits
