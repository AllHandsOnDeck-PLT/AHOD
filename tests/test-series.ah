series<int> a
main:
  {
    a = [1]
  	do PRINT(1)
  }
  