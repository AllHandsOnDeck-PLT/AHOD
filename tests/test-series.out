3
apple
1.1
0